### PR TITLE
SQL Server HA/DR Availability Group queries

### PR DESCRIPTION
**Pull request by avinash-nigam**
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

Includes changes for 
- Queries for HA/DR Availability Group for database type - SQL Server (IaaS/VM), these are optional queries by default
- Fixed a bug in SQLServerProperties query

### CODE REVIEW

**Comment by avinash-nigam** [comment]
@denzilribeiro could you please review this?

File: plugins/inputs/sqlserver/azuresqlqueries.go [hunk 1]
```diff
@@ -449,13 +449,19 @@ WITH PerfCounters AS (
 			,'Readahead pages/sec'
 			,'Lazy writes/sec'
 			,'Checkpoint pages/sec'
+			,'Free pages'
+			,'Extension free pages'
```
Comment (denzilribeiro) [inline plugins/inputs/sqlserver/azuresqlqueries.go:453]: Is this really something we want? Very few folks ever use BP extensions

File: plugins/inputs/sqlserver/azuresqlqueries.go [hunk 2]
```diff
@@ -972,13 +981,19 @@ WITH PerfCounters AS (
...
```
Comment (denzilribeiro) [inline plugins/inputs/sqlserver/azuresqlqueries.go:985]: Is this really a need? Aka is only for BP extensions which aren't used much

File: plugins/inputs/sqlserver/sqlserverqueries.go [hunk 3]
```diff
@@ -331,13 +342,19 @@ SELECT DISTINCT
 			,'Readahead pages/sec'
 			,'Lazy writes/sec'
 			,'Checkpoint pages/sec'
+			,'Free pages'
+			,'Extension free pages'
```
Comment (denzilribeiro) [inline plugins/inputs/sqlserver/sqlserverqueries.go:346]: Is this really a need? Aka is only for BP extensions which aren't used much

File: plugins/inputs/sqlserver/sqlserver.go [hunk 4]
```diff
@@ -15,14 +15,15 @@ import (
...
-	ExcludeQuery  []string `toml:"exclude_query"`
-	queries       MapQuery
-	isInitialized bool
+	Servers             []string `toml:"servers"`
+	QueryVersion        int      `toml:"query_version"`
+	AzureDB             bool     `toml:"azuredb"`
+	DatabaseType        string   `toml:"database_type"`
+	IncludeQuery        []string `toml:"include_query"`
+	ExcludeQuery        []string `toml:"exclude_query"`
+	IgnoreDefaultServer bool     `toml:"ignore_default_server"`
```
Comment (denzilribeiro) [inline plugins/inputs/sqlserver/sqlserver.go:24]: This should be a separate PR or else will get lost in the HADR changes, this has nothing to do with HADR right?

**Review by denzilribeiro — CHANGES_REQUESTED** [review]
Ideally please split the PR as it has HADR changes, a config change and Perfmon counter fixes.. Limit this one to HADR related, and then create a new PR for other stuff so is easily trackable.

**Comment by avinash-nigam** [comment]
> Ideally please split the PR as it has HADR changes, a config change and Perfmon counter fixes.. Limit this one to HADR related, and then create a new PR for other stuff so is easily trackable.

Pulled out the perfmon counters' commit from this PR. Now this only has HADR related changes.

File: plugins/inputs/sqlserver/sqlserver.go (hunk 4 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go:24]: Makes sense, pulled out last commit to only have HADR changes.

File: plugins/inputs/sqlserver/azuresqlqueries.go (hunk 1 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/azuresqlqueries.go]: This was based on a request. However, since these changes have been pulled out of this PR, I will address this offline or in the next PR.

File: plugins/inputs/sqlserver/azuresqlqueries.go (hunk 2 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/azuresqlqueries.go]: This was based on a request. However, since these changes have been pulled out of this PR, I will address this offline or in the next PR.

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 3 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserverqueries.go]: This was based on a request. However, since these changes have been pulled out of this PR, I will address this offline or in the next PR.

File: plugins/inputs/sqlserver/sqlserverqueries.go [hunk 5]
```diff
@@ -1150,3 +1150,91 @@ FROM (
 	ORDER BY [record_id] DESC
 ) AS z
 `
+
+const sqlServerAvailabilityReplicaStates string = `
+IF SERVERPROPERTY('EngineEdition') NOT IN (2,3,4) BEGIN /*NOT IN Standard,Enterpris,Express*/
```
Comment (denzilribeiro) [inline plugins/inputs/sqlserver/sqlserverqueries.go:1155]: Can you also add this check - aka collect only if availability groups enabled

IF SERVERPROPERTY('EngineEdition') NOT IN (2,3,4)  or SERVERPROPERTY('IsHadrEnabled') = 0

File: plugins/inputs/sqlserver/sqlserverqueries.go [hunk 6]
```diff
@@ -1150,3 +1150,91 @@ FROM (
...
+	last_connect_error_number,
+	last_connect_error_description,
+	last_connect_error_timestamp
+from sys.dm_hadr_availability_replica_states AS hars
+inner join sys.availability_replicas AS ar on hars.replica_id = ar.replica_id
+inner join sys.availability_groups AS ag on ar.group_id = ag.group_id
+`
+
+const sqlServerDatabaseReplicaStates string = `
+IF SERVERPROPERTY('EngineEdition') NOT IN (2,3,4) BEGIN /*NOT IN Standard,Enterpris,Express*/
```
Comment (denzilribeiro) [inline plugins/inputs/sqlserver/sqlserverqueries.go:1202]: Can you also add this check? only if availbility groups enabled
IF SERVERPROPERTY('EngineEdition') NOT IN (2,3,4)  or SERVERPROPERTY('IsHadrEnabled') = 0

File: plugins/inputs/sqlserver/sqlserver.go [hunk 7]
```diff
@@ -46,62 +47,62 @@ const sampleConfig = `
...
+## SQLServerSchedulers, SQLServerRequests, SQLServerVolumeSpace, SQLServerCpu
+
+database_type = "SQLServer"
+
+include_query = []
+
+## SQLServerAvailabilityReplicaStates and SQLServerDatabaseReplicaStates are optional queries and hence excluded here as default
+exclude_query = ["SQLServerAvailabilityReplicaStates", "SQLServerDatabaseReplicaStates"]
+
+## Set this flag to true only when you don't want to use the default server used for collecting SQL Server (database_type = "SQLServer") metrics.
```
Comment (denzilribeiro) [inline plugins/inputs/sqlserver/sqlserver.go:91]: Worth mentioning if you don't have a connection string defined we will connect to default "local" instance which isn't relevant for Azure.

**Review by denzilribeiro — CHANGES_REQUESTED** [review]
Looks good, other than 2 checks requested

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 5 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserverqueries.go]: Fixed !

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 6 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserverqueries.go]: Fixed!

File: plugins/inputs/sqlserver/sqlserver.go (hunk 7 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: Made the comment more explicit

**Review by denzilribeiro — CHANGES_REQUESTED** [review]
Avinash I am super sorry about this was my mistake. I had asked to add this check
IF SERVERPROPERTY('EngineEdition') NOT IN (2,3,4) OR SERVERPROPERTY('IsHadrEnabled')
But now am thinking if someone doesn't have HADR enabled, the log will be flooded with messages.
I think would be better to revert this way. My apologies for having you make the changes that I didn't think out well.


IF SERVERPROPERTY('EngineEdition') NOT IN (2,3,4) 
... Error message

IF SERVERPROPERTY('IsHadrEnabled') ==1
.. Execute query

File: plugins/inputs/sqlserver/sqlserverqueries.go [hunk 8]
```diff
@@ -1150,3 +1150,91 @@ FROM (
...
+	last_redone_time,
+	log_send_queue_size,
+	log_send_rate,
+	redo_queue_size,
+	redo_rate,
+	filestream_send_rate,
+	last_commit_time,
+	secondary_lag_seconds
+from sys.dm_hadr_database_replica_states AS drs
+inner join sys.availability_replicas AS ar on drs.replica_id = ar.replica_id
```
Comment (mikedavem) [inline plugins/inputs/sqlserver/sqlserverqueries.go:1239]: Hi, may I suggest to add information of availability group state (or maybe a additional query ) ?

Something as follows:

`SELECT
	'sqlserver_hadr_avreplica_states' AS [measurement],
	REPLACE(@@SERVERNAME,'\',':') AS [sql_instance],
	convert(nvarchar(36), hars.replica_id) as replica_id,
	ar.replica_server_name,
	convert(nvarchar(36), hars.group_id) as group_id,
	ag.name AS group_name,
	ag.basic_features,
	ag.is_distributed,
	hags.synchronization_health_desc AS ag_synchronization_health_desc,
	ar.replica_metadata_id,
	ar.availability_mode ,
	ar.availability_mode_desc,
	ar.failover_mode,
	ar.failover_mode_desc,
	ar.session_timeout,
	ar.primary_role_allow_connections,
	ar.primary_role_allow_connections_desc,
	ar.secondary_role_allow_connections,
	ar.secondary_role_allow_connections_desc,
	ar.seeding_mode,
	ar.seeding_mode_desc,
	hars.is_local,
	hars.role,
	hars.role_desc,
	hars.operational_state,
	hars.operational_state_desc,
	hars.connected_state,
	hars.connected_state_desc,
	hars.recovery_health,
	hars.recovery_health_desc,
	hars.synchronization_health AS replica_synchronization_health,
	hars.synchronization_health_desc AS replica_synchronization_health_desc,
	hars.last_connect_error_number,
	hars.last_connect_error_description,
	hars.last_connect_error_timestamp
from sys.dm_hadr_availability_replica_states AS hars
inner join sys.availability_replicas AS ar on hars.replica_id = ar.replica_id
inner join sys.availability_groups AS ag on ar.group_id = ag.group_id
INNER JOIN sys.dm_hadr_availability_group_states AS hags ON hags.group_id = ag.group_id`

File: plugins/inputs/sqlserver/sqlserverqueries.go [hunk 9]
```diff
@@ -1150,3 +1150,91 @@ FROM (
...
+	recovery_health,
+	recovery_health_desc,
+	synchronization_health,
+	synchronization_health_desc,
+	last_connect_error_number,
+	last_connect_error_description,
+	last_connect_error_timestamp
+from sys.dm_hadr_availability_replica_states AS hars
+inner join sys.availability_replicas AS ar on hars.replica_id = ar.replica_id
+inner join sys.availability_groups AS ag on ar.group_id = ag.group_id
```
Comment (mikedavem) [inline plugins/inputs/sqlserver/sqlserverqueries.go:1198]: Hi, very nice idea :) May I suggest to add info about AG state in this query (or maybe in a different query to keep consistent) ? :

```
SELECT
	'sqlserver_hadr_avreplica_states' AS [measurement],
	REPLACE(@@SERVERNAME,'\',':') AS [sql_instance],
	convert(nvarchar(36), hars.replica_id) as replica_id,
	ar.replica_server_name,
	convert(nvarchar(36), hars.group_id) as group_id,
	ag.name AS group_name,
	ag.basic_features,
	ag.is_distributed,
	hags.synchronization_health_desc AS ag_synchronization_health_desc,
	ar.replica_metadata_id,
	ar.availability_mode ,
	ar.availability_mode_desc,
	ar.failover_mode,
	ar.failover_mode_desc,
	ar.session_timeout,
	ar.primary_role_allow_connections,
	ar.primary_role_allow_connections_desc,
	ar.secondary_role_allow_connections,
	ar.secondary_role_allow_connections_desc,
	ar.seeding_mode,
	ar.seeding_mode_desc,
	hars.is_local,
	hars.role,
	hars.role_desc,
	hars.operational_state,
	hars.operational_state_desc,
	hars.connected_state,
	hars.connected_state_desc,
	hars.recovery_health,
	hars.recovery_health_desc,
	hars.synchronization_health AS replica_synchronization_health,
	hars.synchronization_health_desc AS replica_synchronization_health_desc,
	hars.last_connect_error_number,
	hars.last_connect_error_description,
	hars.last_connect_error_timestamp
from sys.dm_hadr_availability_replica_states AS hars
inner join sys.availability_replicas AS ar on hars.replica_id = ar.replica_id
inner join sys.availability_groups AS ag on ar.group_id = ag.group_id
INNER JOIN sys.dm_hadr_availability_group_states AS hags ON hags.group_id = ag.group_id
```

Comment (denzilribeiro) [reply plugins/inputs/sqlserver/sqlserverqueries.go:1198]: I am not for having a separate query, which means yet another one to run in an iteration. @avinash-nigam  if adding group state you may have to name the columns appropriately for state :)

One more thing @avinash-nigam  - can we rename the measurement to sqlserver_hadr_replica_states ( no need for the av part?)

Thanks much!

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 5 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserverqueries.go:1155]: reverted this change and did the IsHadrEnabled check before executing the query

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 6 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserverqueries.go:1206]: reverted this change and did the IsHadrEnabled check before executing the query

**Comment by avinash-nigam** [comment]
> Avinash I am super sorry about this was my mistake. I had asked to add this check
> IF SERVERPROPERTY('EngineEdition') NOT IN (2,3,4) OR SERVERPROPERTY('IsHadrEnabled')
> But now am thinking if someone doesn't have HADR enabled, the log will be flooded with messages.
> I think would be better to revert this way. My apologies for having you make the changes that I didn't think out well.
> 
> IF SERVERPROPERTY('EngineEdition') NOT IN (2,3,4)
> ... Error message
> 
> IF SERVERPROPERTY('IsHadrEnabled') ==1
> .. Execute query

@denzilribeiro made changes as per your suggestion

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 9 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserverqueries.go]: @mikedavem included AG state as per your and Denzil's suggestions
@denzilribeiro removed the "av" part from the measurement

**Review by denzilribeiro — APPROVED** [review]
Looks good @Trovalo  - can you also take a look at this?

**Comment by avinash-nigam** [comment]
> Looks good @Trovalo - can you also take a look at this?

@Trovalo - any thoughts on the PR? or, shall I go ahead and request for a merge?

**Review by Trovalo — APPROVED** [review]
Sorry I got overwhelmed and forgot about it the last week. Looks fine to me.

**Comment by avinash-nigam** [comment]
Thank you @denzilribeiro and @Trovalo for the review and approvals.
@ssoroka and @reimda - could you please help merge the changes?

File: plugins/inputs/sqlserver/sqlserver.go (hunk 4 above)
Comment (ssoroka) [inline plugins/inputs/sqlserver/sqlserver.go:24]: I don't understand why we need an "ignore default server" flag.

Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go:24]: Thank you @ssoroka for taking a look at this. The "ignore default server" flag was added for the case when the server list is empty and no server should be monitored. Without the flag, an empty server list would cause the localhost to be monitored, which would throw errors when using the agent for remote monitoring. Also, the localhost mechanism is valid only for SQL Server database type, but not for Azure SQL Database or Azure SQL Managed Instance. By adding the flag, the previous experience would work as is, and whoever does not want to monitor localhost can use this flag.

Comment (Trovalo) [reply plugins/inputs/sqlserver/sqlserver.go:24]: I didn't spot that. tbh, I don't get why someone should leave the list empty even for localhost, I know the plugin will look for the default instance (port 1433) on localhost but I've never used that anyway as I want all to be explicitly set.

I'm not against the change, I just find it odd, if you don't want to use the plugin you can simply comment it out, or if you use multiple .conf file and the `--config-directory` you can simply change the file extension to something different than .conf and it will be ignored (I personally use this pattern `input_sqlserver.conf.ignore`.

In any case, this looks out of scope for this PR, as it should just add the new queries, can you propose this in a new  PR?

Comment (ssoroka) [reply plugins/inputs/sqlserver/sqlserver.go:24]: This isn't how we normally do defaults anyway. the default should go in the init() function, and then you don't need conditional checks like 
```

	if len(s.Servers) == 0 {
		s.Servers = append(s.Servers, defaultServer)
	}
```
that assume empty lists are unset. Let's remove this boolean and move the default to the init() function, and then explicitly empty lists will work properly.

Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go:24]: Thank you @ssoroka and @Trovalo for the feedback. Let me pull this change out and create a separate PR for it, so that HA/DR queries related changes can go ahead and get merged.

Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: @ssoroka and @Trovalo - I've reverted the default flag related changes and will create a separate PR for it. This PR now contains only changes for HA/DR queries. Please let me know if you have further feedback on this. If not, it'd be great to get this merged. Thank you.

**Review by denzilribeiro — APPROVED** [review]

File: plugins/inputs/sqlserver/sqlserver.go (hunk 4 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: @ssoroka - Could you please help merge this PR if there are no concerns?

File: plugins/inputs/sqlserver/sqlserver.go [hunk 10]
```diff
@@ -227,7 +221,7 @@ func (s *SQLServer) Gather(acc telegraf.Accumulator) error {
 		}
 	}
 
-	if len(s.Servers) == 0 && s.DatabaseType == "SQLServer" && !s.IgnoreDefaultServer {
+	if len(s.Servers) == 0 {
 		s.Servers = append(s.Servers, defaultServer)
 	}
```
Comment (ssoroka) [inline plugins/inputs/sqlserver/sqlserver.go]: This needs to move to the init() function to be override-able. You won't need the conditional there, so just s.Servers = []string{defaultServer}

Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: @ssoroka - That's an existing code, I can address it when I try to address the default server aspect in a separate PR. This PR is only for the HA/DR queries now.

Comment (ssoroka) [reply plugins/inputs/sqlserver/sqlserver.go]: Alright. Please do make another PR for it, as it's currently busted, as you've noticed.

Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: Sure, will do that. Thank you!
